File: treebylevels_opt.py
from collections import deque

class Node:
    def __init__(self, L, R, n):
        self.left = L
        self.right = R
        self.value = n
        

def tree_by_levels(node):
    if not node:
        return []
    
    res, queue = [], deque([node])
    
    while queue:
        n = queue.popleft()
        res.append(n.value)
        queue.extend(child for child in (n.left, n.right) if child)
    
    return res


tree_by_levels(None)

tree_by_levels(Node(Node(None, Node(None, None, 4), 2), Node(Node(None, None, 5), Node(None, None, 6), 3), 1))

File: treebylevels_opt.js
class Node { 
  constructor(value, left = null, right = null) {
    this.value = value;
    this.left  = left;
    this.right = right;
  }
}

function treeByLevels(rootNode) {
  if (!rootNode) return [];

  const result = [];
  const queue = [rootNode];

  for (let i = 0; i < queue.length; i++) {
    const node = queue[i];
    result.push(node.value);
    node.left && queue.push(node.left);
    node.right && queue.push(node.right);
  }

  return result;
}

treeByLevels(null)

const treeOne =
  new Node(2,
    new Node(8,
      new Node(1),
      new Node(3)
    ),
    new Node(9,
      new Node(4),
      new Node(5)
    )
  );

treeByLevels(treeOne)

const treeTwo =
  new Node(1,
    new Node(8,
      null,
      new Node(3)
    ),
    new Node(4,
      null,
      new Node(5,
        null,
        new Node(7)
      )
    )
);

treeByLevels(treeTwo)